proposals/0000-mutable-fields.rst

```diff
@@ -0,0 +1,335 @@
+
+  readRef :: Mutable a -> IO a
+  readRef aref = IO $ \s -> readRef# aref
+
+But what *are* these ``Ref#`` things that are extracted from the
+constructor by pattern matching and seemingly passed to ``readRef#``?
+The idea is that the runtime representation of ``Ref#`` is a pair of
+the containing object and the offset of the mutable field. Just before
+code generation, probably in the Unarise phase, we will manifest the
+runtime representation of ``Ref#`` at each pattern match::
+
+  readMutPair = \m ->
+    case m of (v::Any) { MutPair aref bref ->
+    let aref = (# v, 0# #) in
+    let bref = (# v, 1# #) in
+    readRef aref >>= \a ->
+    readRef bref >>= \b ->
+    return (a,b) }
+
+and then propagate the expansion of ``aref`` and ``bref`` to all the
+places they are referenced. Function arguments of type ``Ref#`` are
+expanded to unboxed pairs of type ``(# Any, Int# #)``.
+
+The ``readRef#`` primitive would be compiled inline to a single read
+instruction. Similarly ``writeRef#`` would compile to a primitive write
+instruction, but it would also need a memory barrier just like
+``writeMutVar#``, and a GC write barrier (the equivalent of
+``dirty_MUT_VAR()``).
+
+
+Garbage collection
+^^^^^^^^^^^^^^^^^^
+
+The garbage collector needs to know that an object is mutable, and
+which fields are mutable.  So we have to put this information in the
+info table.  Probably:
+
+- new ``MUT_CONSTR_CLEAN`` and ``MUT_CONSTR_DIRTY`` object types
+
+- Each constructor with a mutable field needs two info tables: the
+  CLEAN one and the DIRTY one.  From each info table we need to be
+  able to get both CLEAN and DIRTY info pointers, for the GC write
+  barrier.
+
+- Representation: put the mutable fields first, before the non-mutable
+  pointers.
+
+- Store the number of mutable fields in the info table, alongside the
+  number of pointers and non-pointers.  (but include the mutable
+  fields in the count of pointer fields, so that RTS code that doesn't
+  care about mutability can work without changes)
+
+The GC needs to do the same CLEAN/DIRTY and non-eager promotion stuff
+that it does with other mutable objects.
+
+TODO: pin down the details of info table representation and the form
+of the GC write barrier code.
+
+Drawbacks
+---------
+
+The GC write barrier for a mutable constructor may be a little less
+efficient than the write barrier for a ``MutVar#``, but this is more
+than compensated for by losing a layer of indirection.
+
+Adding new syntax has lots of costs: changes in the parser and
+`HsSyn`, the renamer and typechecker, not to mention knock-on effects
+on external packages: `haskell-src-exts` and clients of that.
+
+Types that contain mutable fields cannot be UNPACKed into other
+constructors.
+
+Alternatives
+------------
+
+Don't do this :)
+
+Unresolved Questions
+--------------------
+
+* Should we provide a way to test for reference equality between
+  mutable constructors?
```

Comment:
Good point.  So deriving `Eq` should work on a type with mutable fields and give you reference equality.  But what if the type has multiple constructors, and some of them are pure?  Then I suspect we would have to disallow deriving `Eq`.

Also if you want `Ord` you have to implement it yourself, with an atomic counter to get a thread-safe total ordering for example.